### DISCUSSION TITLE
Forcing constants to be dimensionless

### DISCUSSION THREAD

**Discussion by Kacpiolo** [MilesCranmer/PySR#606] [Q&A] [answered]
I am using the dimensional constraints in my PySR model.

For context here are units of the dataset that I am testing with:
```
X_units = ["m/s", "kg/m^4", "kg/ms^2", "", "s", "m/s", "m^2/s", "s^-1", "m^-1", "s^-1"]
y_units = ""
```

Question 1. I just would like to double check that I can set a dimensionless units with "".

Question 2. When I am fitting my current model (dimensional_constraint_penalty=10**5) the units of all constants are [.]. I know that in my field, in this particular case, we would expect these constants to be dimensionless as this forces the physical meaning of the equation to come from the variables being used and their relation. I would like to put a dimensionless unit constraint on coefficients in order to force PySR to create physically meaningful relations. How should I go about doing that?

Thank you in advance.

**Comment by MilesCranmer** [comment]
> Question 1. I just would like to double check that I can set a dimensionless units with "".

Correct although I might prefer using `"1"` as its more explicit.

> Question 2. When I am fitting my current model (dimensional_constraint_penalty=10**5) the units of all constants are [.]. I know that in my field, in this particular case, we would expect these constants to be dimensionless as this forces the physical meaning of the equation to come from the variables being used and their relation. I would like to put a dimensionless unit constraint on coefficients in order to force PySR to create physically meaningful relations. How should I go about doing that?

When it says the units are `[⋅]` that simply means they are unknown to the algorithm. (Maybe PySR should show it as a `[?]` instead...)

The reason for this is that it's expensive to calculate exact units for every constant in an expression (and sometimes there are multiple valid solutions!). However, you can verify if an expression is dimensionally _possible_ quite cheaply.

See https://astroautomata.com/PySR/examples/#10-dimensional-constraints for info on this.

<img width="1073" alt="Screenshot 2024-04-27 at 18 00 24" src="https://github.com/MilesCranmer/PySR/assets/7593028/a0501300-0d80-4c2a-9ac1-8e7849f9e48b">

**Reply by Kacpiolo** [reply]
Hello @MilesCranmer, thank you for the reply.

I'm afraid that my original point still remains. Is there a way to set all constant units to always be dimensionless?

I am using PySR as physics discovery in a way but, for the equation to have a physical meaning it must be dimensionally consistent and the physical meaning must come from the features being used, not constants. I.e. the consultant must always be dimensionless.

Anything other than unitless constant won't be accepted in a top journal, which is why this is so important to me. This also seems to be the last hurdle for us to get over before running expensive computations.

**Reply by MilesCranmer** [reply]
Oh, I see what you mean. That is not yet implemented but I can try to add it.

**Reply by MilesCranmer** [reply]
Ok I added it https://github.com/MilesCranmer/SymbolicRegression.jl/pull/310 under the parameter `dimensionless_constants_only`.

Not in PySR yet but will be soon.

**Comment by MilesCranmer** [comment]
Edit: made https://github.com/MilesCranmer/SymbolicRegression.jl/pull/307 which switches from `[⋅]` to `[?]` which is hopefully less confusing. Wdyt @Kacpiolo?

**Comment by MilesCranmer** [comment, marked as answer]
I added this feature in #608 with the `dimensionless_constants_only` parameter (set to `True` for your use-case).

I will merge this in a few days once the global registries are safely up-to-date with the new SymbolicRegression backend.

In the meantime you can install it yourself with, e.g.,

```bash
pip install git+https://github.com/MilesCranmer/PySR.git@dimensionless-constants
```

**Edit:** Is available in PySR 0.18.4.

**Reply by Kacpiolo** [reply]
I have tested the dimensionless_constants_only and it works well. Thank you so much!
